src/scheduler/pbs_bitmap.c

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright (C) 1994-2018 Altair Engineering, Inc.
+ * For more information, contact Altair at www.altair.com.
+ *
+ * This file is part of the PBS Professional ("PBS Pro") software.
+ *
+ * Open Source License Information:
+ *
+ * PBS Pro is free software. You can redistribute it and/or modify it under the
+ * terms of the GNU Affero General Public License as published by the Free
+ * Software Foundation, either version 3 of the License, or (at your option) any
+ * later version.
+ *
+ * PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Commercial License Information:
+ *
+ * For a copy of the commercial license terms and conditions,
+ * go to: (http://www.pbspro.com/UserArea/agreement.html)
+ * or contact the Altair Legal Department.
+ *
+ * Altair’s dual-license business model allows companies, individuals, and
+ * organizations to create proprietary derivative works of PBS Pro and
+ * distribute them - whether embedded or bundled with other software -
+ * under a commercial license agreement.
+ *
+ * Use of Altair’s trademarks, including but not limited to "PBS™",
+ * "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+ * trademark licensing policies.
+ *
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "pbs_bitmap.h"
+
+#define BYTES_TO_BITS(x) ((x) * 8)
+
+
+/**
+ * @brief allocate space for a pbs_bitmap (and possibly the bitmap itself)
+ * @param pbm - bitmap to allocate space for.  NULL to allocate a new bitmap
+ * @param num_bits - number of bits to allocate
+ * @return pbs_bitmap *
+ * @retval bitmap which was allocated
+ * @retval NULL on error
+ */
+pbs_bitmap *
+pbs_bitmap_alloc(pbs_bitmap *pbm, long num_bits)
+{
+	pbs_bitmap *bm;
+	unsigned long *tmp_bits;
+	long prev_longs;
+	
+	if(num_bits <= 0)
+		return NULL;
+	
+	if(pbm == NULL) {
+		bm = malloc(sizeof(pbs_bitmap));
+		if(bm == NULL)
+			return NULL;
+		bm->bits = NULL;
+		bm->num_longs = 0;
+	}
+	else 
+		bm = pbm;
+	
+	/* If we have enough unused bits available, we don't need to allocate */
+	if (bm->num_longs * BYTES_TO_BITS(sizeof(unsigned long)) >= num_bits) {
+		bm->num_bits = num_bits;
```

Comment:
I understand your confusion.  While num_longs and num_bits are are connected, they aren't the same.  num_longs is the amount of data allocated to the bitmap.  num_bits is the high order bit.  If num_longs is 2, then there are 128 bits available to the bitmap.  num_bits can be 65.  This means there are 63 bits allocated to the bitmap but not used.  This can get worse if do a pbs_bitmap_equals() to a smaller bitmap.  The amount of data allocated to a bitmap only grows.  It never shrinks.  So this is saying if we have enough memory allocated, then we don't need to allocate any more.  In the above example, we have 128 bits allocated.  If the num_bits parameter is 70, then we don't need to allocate any more memory.  We just set bm->num_bits  to 70.